python/ray/tune/suggest/hebo.py

```diff
@@ -0,0 +1,359 @@
+    Searcher:
+
+    .. code-block:: python
+
+        from ray import tune
+        from ray.tune.suggest.hebo import HEBOSearch
+        from bo.design_space.design_space import DesignSpace
+
+        space_config = [
+            {'name' : 'width', 'type' : 'num', 'lb' : 0, 'ub' : 20},
+            {'name' : 'height', 'type' : 'num', 'lb' : -100, 'ub' : 100},
+        ]
+        space = DesignSpace().parse(space_config)
+
+        hebo = HEBOSearch(space, metric="mean_loss", mode="min")
+        tune.run(my_func, search_alg=hebo)
+
+    """
+
+    def __init__(
+            self,
+            space: Optional[Union[
+                Dict, "hebo.design_space.design_space.DesignSpace"]] = None,
+            metric: Optional[str] = None,
+            mode: Optional[str] = None,
+            points_to_evaluate: Optional[List[Dict]] = None,
+            evaluated_rewards: Optional[List] = None,
+            random_state_seed: Optional[int] = None,
+            **kwargs):
+        assert hebo is not None, (
+            "HEBO must be installed!. You can install HEBO with"
+            " the command: `pip install "
+            "git+https://github.com/huawei-noah/noah-research.git"
+            "#subdirectory=HEBO`.")
+        if mode:
+            assert mode in ["min", "max"], "`mode` must be 'min' or 'max'."
+        if random_state_seed:
+            assert random_state_seed is None or isinstance(
+                random_state_seed, int
+            ), "random_state_seed must be None or int, got '{}'.".format(
+                type(random_state_seed))
+        super(HEBOSearch, self).__init__(metric=metric, mode=mode)
+
+        if isinstance(space, dict) and space:
+            resolved_vars, domain_vars, grid_vars = parse_spec_vars(space)
+            if resolved_vars:
+                raise TypeError(SPACE_ERROR_MESSAGE)
+            if domain_vars or grid_vars:
+                logger.warning(
+                    UNRESOLVED_SEARCH_SPACE.format(
+                        par="space", cls=type(self)))
+                space = self.convert_search_space(space)
+        elif space is not None and not isinstance(
+                space, hebo.design_space.design_space.DesignSpace):
+            raise TypeError(SPACE_ERROR_MESSAGE +
+                            " Got {}.".format(type(space)))
+
+        self._hebo_config = kwargs
+        self._random_state_seed = random_state_seed
+        self._space = space
+        self._points_to_evaluate = points_to_evaluate
+        self._evaluated_rewards = evaluated_rewards
+        self._initial_points = []
+        self._live_trial_mapping = {}
+
+        self._opt = None
+        if space:
+            self._setup_optimizer()
+
+    def _setup_optimizer(self):
+        # HEBO internally minimizes, so "max" => -1
+        if self._mode == "max":
+            self._metric_op = -1.
+        elif self._mode == "min":
+            self._metric_op = 1.
+
```

Comment:
We should check here that `self._space` is not empty (apparently this check seems to be missing from some other searchers?)

Reply:
I copied it from one of the existing ones, so it would appear so. I'll add it here, though the rest should probably be fixed in another PR.

Reply:
@krfricke Should it throw an exception if `self._space` is empty?

Reply:
Yes - it's enough to do it for this Searcher here. We'll tackle the other ones somewhere else